ssl/statem/statem_clnt.c

```diff
@@ -584,6 +587,7 @@ WRITE_TRAN ossl_statem_client_write_transition(SSL_CONNECTION *s)
          * No transition at the end of writing because we don't know what
          * we will be sent
          */
+        s->ts_msg_write = ossl_time_now();
```

Comment:
To be clear, we do need to have this assignment to `s->ts_msg_write` for the non-early-data case.  But when we enter this (`TLS_ST_CW_CLNT_HELLO`) block there is a path where we can leave the block without ever storing the first timestamp -- the case when `s->early_data_state == SSL_EARLY_DATA_CONNECTING`.  In that case we will continue to send and go through this state machine again before we're done sending (`WRITE_TRAN_FINISHED`).  So my proposal is to leave this line as-is, and add an additional assignment between lines 605 and 606 (RHS of diff), so that we set the timestamp when we're transitioning out of `TLS_ST_EARLY_DATA`.  I believe that will match the documented behavior

Reply:
I've added a timestamp capture between lines 605 and 606 as well as created a 5th test for client-side early data